src/token.ts

```diff
@@ -1,8 +1,393 @@
+}
+
+export class TransitionToken implements Token {
+    readonly type = 'transition'
+    readonly text: string
+
+    constructor(line: string) {
+        const match = line.match(regex.transition)
+        this.text = match[1] || match[2]
+    }
+
+    addTo(tokens: Token[]): Token[] {
+        return [...tokens, this]
+    }
+
+    static matchedBy(line: string): boolean {
+        return regex.transition.test(line)
+    }
+}
+
+export class DialogueBlock implements Block {
+    readonly tokens: Token[] = []
+    readonly dual: boolean
+    readonly too_short: boolean
+    
+    constructor(line: string, dual: boolean) {
+        const match = line.match(regex.dialogue);
+
+        let name = match[1] || match[2];
+        this.too_short = isTooShort(name) && !line.startsWith('@')
+
+        // iterating from the bottom up, so push dialogue blocks in reverse order
+        const isDualDialogue = !!(match[3]);
+        if (isDualDialogue) {
+            this.tokens.push(new DualDialogueEndToken());
+        }
+
+        this.tokens.push(new DialogueEndToken());
+
+        const parts: string[] = match[4].split(/(\(.+\))(?:\n+)/).reverse();
+        this.tokens.push(...parts.reduce((p, text = '') => {
+            if (!text.length) {
+                return p
+            }
+            if (regex.parenthetical.test(text)) {
+                return [...p, new ParentheticalToken(text)]
+            }
+            if (regex.lyrics.test(text)) {
+                return [...p, new LyricsToken(text)]
+            }
+            return [...p, new DialogueToken(text)]
+        }, []))
+
+        this.tokens.push(
+            new CharacterToken(name.trim()),
+            new DialogueBeginToken(
+                isDualDialogue ? 'right' : dual ? 'left' : undefined
+            )
+        );
+
+        if (dual) {
+            this.tokens.push(new DualDialogueBeginToken());
+        }
+
+        this.dual = isDualDialogue;
+    }
+
+    addTo(tokens: Token[]): Token[] {
+        return [...tokens, ...this.tokens]
+    }
+
+    static matchedBy(line: string): boolean {
+        return regex.dialogue.test(line)
+    }
+}
+
+export class DialogueBeginToken implements Token {
+    readonly type = 'dialogue_begin'
+    readonly dual: 'left' | 'right' | undefined
+
+    constructor(dual?: 'left' | 'right') {
+        this.dual = dual
+    }
+
+    addTo(tokens: Token[]): Token[] {
+        return [...tokens, this]
+    }
+}
+
+export class CharacterToken implements Token {
+    readonly type = 'character'
+    readonly text: string
+
+    constructor(text: string) {
+        this.text = text
+    }
+
+    addTo(tokens: Token[]): Token[] {
+        return [...tokens, this]
+    }
+}
+
+export class DialogueToken implements Token {
+    readonly type = 'dialogue'
+    readonly text: string
+    readonly scene_number: string
+
+    constructor(text: string, scene_number?: string) {
+        this.text = text
+        this.scene_number = scene_number
+    }
```

Comment:
I don't believe a scene number for dialogue is in the [Fountain spec](https://fountain.io/syntax/#scene-headings), none at least that I can see.

Is it fair to assume this was copied from another method? Adding to this, it seems like the `scene_number` parameter is never used in the `new DialogueToken()` on line 152.

 ```suggestion
    constructor(text: string) {
        this.text = text
    }
```